src/main/java/vct/col/veymont/TerminationCheck.scala

```diff
@@ -0,0 +1,199 @@
+      case Some(tup) => {
+        val method = tup._1
+        val mClass = tup._2
+        if(mClass == mainClassName && currentClass != mainClassName) {
+          Fail("VeyMont Fail: Cannot call Main method '%s' from role or other class! %s", m.method,m.getOrigin)
+        } else { //if (mClass != mainClassName || method.kind == Method.Kind.Pure) {
+          methodCalled = true
+          encountered += MethodInvToString(m)
+          visit(method)
+        // } else {
+          //stop checking; recursion is allowed in non-pure main methods
+        }
+      }
+      case None => {
+        if(!(m.method == Method.JavaConstructor && currentClass == mainClassName)) //Main constructor is fine, so skip
+          Fail("VeyMont Fail: Method invocation %s not allowed! %s", m.method, m.getOrigin)
+      }
+    }
+  }
+
+  def MethodInvToString(mi : MethodInvokation) = mi.method //(if (mi.`object`==null) "" else mi.`object`.toString) + mi.method
+
+  override def visit(l : LoopStatement) : Unit =
+    if(currentClass != mainClassName)
+      Warning("Session Warning: loop in method of non-Main class" + deadlockWarning)
+
+  override def visit(s : ASTSpecial) : Unit = {
+    s.kind match {
+      case ASTSpecial.Kind.Goto => Warning("Session Warning: Goto" + deadlockWarning)
+      case ASTSpecial.Kind.Wait => Warning("Session Warning: Wait" + deadlockWarning)
+      case ASTSpecial.Kind.Fork => Warning("Session Warning: Fork" + deadlockWarning)
+      case ASTSpecial.Kind.Join => Warning("Session Warning: Join" + deadlockWarning)
+      case ASTSpecial.Kind.Lock => Warning("Session Warning: Lock" + deadlockWarning)
+      case ASTSpecial.Kind.Unlock => Warning("Session Warning: Unlock" + deadlockWarning)
+      case ASTSpecial.Kind.Send => Warning("Session Warning: Send" + deadlockWarning)
+      case ASTSpecial.Kind.Recv => Warning("Session Warning: Recv" + deadlockWarning)
+      case _ => super.visit(s)
+    }
+  }
+/*
+  private def isForLoop(l : LoopStatement) : Boolean = {
+    val guardOp = l.getEntryGuard.asInstanceOf[OperatorExpression]
+    val initval = l.getInitBlock.asInstanceOf[BlockStatement].getStatement(0).asInstanceOf[VariableDeclaration].get().head.asInstanceOf[DeclarationStatement].init.get.asInstanceOf[ConstantExpression]
+
+    val updateOp = l.getUpdateBlock.asInstanceOf[BlockStatement].getStatement(0)
+    //TODO: get Value from initval
+    if(isLengthExpr(guardOp.arg(1)))
+      initval.value == 0 && Set(StandardOperator.LT).contains(guardOp.operator) && updateOp.operator == StandardOperator.PostIncr
+    else initval.value < guardOp.arg(1).asInstanceOf[ConstantExpression].value && updateOp.operator == StandardOperator.PostIncr ||
+      initval.value > guardOp.arg(1).asInstanceOf[ConstantExpression].value && updateOp.operator == StandardOperator.PostDecr
+  }
+*/
+  private def isLengthExpr(e : ASTNode) : Boolean = e match {
+    case d : Dereference => d.field == Dereference.ArrayLength
```

Comment:
*and* d.object is an array type!

Reply:
I removed this method as I'm not using it.